### ISSUE TITLE
Concepts with inheritance causing 'SIGSSEV: Illegal storage access ...'

### ISSUE THREAD

**Issue by deech** [nim-lang/Nim#10506]
Subtypes don't seem to handle subtypes very well, the following code:
```
type
   Foo = object of RootObj
   Bar = object of Foo

   FooOrBar = concept s, type T
     s is T
     f(s) is string

proc f(f: Foo): string = "Foo"

proc conceptf(f : FooOrBar): string =
  f(f)

let bar = Bar()
echo conceptf(bar)
```
generates the following error:
```
Hint: system [Processing]
Hint: concepttest [Processing]
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```

While Nim should not `SIGSEGV` when compiling this snippet I think concepts should not try and follow inheritance hierarchies and the line `s is T` should mean `s is exactly T` and not `s is T or a subtype of T`. Coming from Scala my experience is subtype + typeclasses (or concepts) causes horrendous type inference and makes concept resolution very hard to reason about.

TLDR: This compiler should not SIGSEGV on this code but it shouldn't typecheck either because concepts + subtyping leads to bad or hard to understand behavior.

**Comment by mratsim** [comment]
T is a compile-time type. At compile-time Bar is Foo, since concepts work at compile-time I don't see a way around your proposition, a normal proc would also accept both Foo and Bar.

**Comment by narimiran** [comment]
Currently (Nim 0.20.99) there is no `SIGSEGV`, but it produces the following: `Error: unhandled exception: types.nim(1126, 10) `b.kind in {tyObject} + skipPtrs`  [AssertionError]`.

**Comment by metagn** [comment]
The call to `inheritanceDiff` which has the assert is [here](https://github.com/nim-lang/Nim/blob/05c74d684481733c507dd18a7da8a296289ca32d/compiler/transf.nim#L599), in transf.
